Migrate a Formula to Multiple Versions: add docs.

MikeMcQuaid opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Add a document describing how to migrate a formula to our new multiple versions support, upgrade a multiple version formula and how to import a homebrew/versions formula into homebrew/core.

CC @Homebrew/maintainers @Homebrew/science @Homebrew/php @Homebrew/linux @alyssais  @sjackman to help me think this through. This is the least disruptive way I can see of having us migrate to the path that will avoid taps needing `revision` bumps immediately in future and instead give them time to upgrade to newer versions while in core we can upgrade en-masse when we choose to while still having the new version be immediately available.

Please feel free not to just pick out flaws but to debate the approach and constraints.

Part of https://github.com/Homebrew/brew/issues/620.

retokromer commented on docs/Migrate-A-Formula-To-Multiple-Versions.md:
```diff
@@ -0,0 +1,27 @@
+
+In separate pull-requests:
+
+1. [Migrate the formula](Migrating-A-Formula-To-A-Tap) from Homebrew/homebrew-versions to Homebrew/homebrew-core with the same, old name e.g. `boost160.rb`.
```
Migrating-A-Formula-To-A-Tap.md

ilovezfs commented:
> while in core we can upgrade en-masse when we choose to while still having the new version be immediately available

Does this mean the new version will get to be non-keg-only right away even if it's not the anointed version with the alias pointing at it yet?

MikeMcQuaid commented:
@DomT4 I'm not disagreeing with your comment but can you comment inline rather than just as a block? There's clearly some overlap between what you're proposing and what I'm proposing but it's hard to see that when they've both been proposed in different formats.

@ilovezfs @DomT4 I realise I've not mentioned `keg_only` in here at all so I'll update it to note that.

DomT4 commented:
> I'm not disagreeing with your comment but can you comment inline rather than just as a block? There's clearly some overlap between what you're proposing and what I'm proposing but it's hard to see that when they've both been proposed in different formats.

To be honest I didn't because although there's overlap it's almost two different mechanisms of going about this entirely. Some of the stuff I proposed and talked about is 100% the opposite of what you've written. It seemed sensible to detail that in context rather than taking a simple blob of what you wrote and writing my different proposal underneath without that broader context.

MikeMcQuaid commented:
@DomT4 I can't evaluate/compare your proposal without seeing how it breaks down to separate pull requests and how we should handle new versions/version upgrades/homebrew-versions migrations. Either inline comments or a new PR following the same basic outline would be appreciated.

DomT4 commented:
Fine, I can file a competing PR. Hadn't done so to avoid wasting one of our's time, but I can submit a competing PR if that's easier for people.

MikeMcQuaid commented:
@DomT4 That'd be great. I'm not convinced my solution is better than yours at all but I'm just struggling to compare them currently.

DomT4 commented:
Will submit later today. Dropped offline last night because painkillers stopped working on the headache 😅.

MikeMcQuaid commented:
> I do think we've overcomplicated this. The idea, from the start really, was that brew install boost gets you the latest boost.

What you seem to be proposing is that it doesn't get you the latest `boost` until we've deleted the current `boost` formula?

> Obviously, that has proven impractical because if you rename the existing boost for example you have to rebuild every single formula that brew uses it, which in many cases is going to lead to enormous CI and user burden. But I think the situation where we rename the current formula and have the alias pointing at the oldest boost is a mistake.

I don't propose the alias points at the oldest boost, just the one that we use by default and we shouldn't use it by default until we've fixed our dependencies such that we can use it. The mass-rebuild in this case would be a one-time thing for the rename whereas it's currently required every time.

> I don't think we need to migrate to the new system overnight, and in fact we get more stability out of things if we don't. As I outlined earlier, using the experience from my work on openssl@1.1 and ILZ's work on boost recently, I think the best process is a gradual move towards deleting the original formula and replacing it at that point with an alias.

I don't think we are proposing moving to it overnight. This proposes we rename one formula to the new format as a test.

> The current boost becomes keg_only. This should require zero or very few immediate revisions because we rewrite linkage to the opt_prefix now.

`brew install boost` shouldn't install a keg-only formula. We should ensure that `brew install boost` creates always `opt/boost`. This should also use https://github.com/Homebrew/brew/pull/1192 or similar to achieve that.

> We add boost@1.62 as a new formula.

We're agreed on this. We can do this today as long as it's keg-only.

> We migrate things over to boost@1.62 as sensibly but quickly as possible.

This migration requires making everything `depends_on "boost@1.6.2"` even for stuff that doesn't break on Boost version updates. This feels like a step back rather than forward to me.

> When brew uses boost returns nothing for all the taps under the Homebrew organisation we delete boost.rb.

Which then breaks all the third-party taps that rely on it working. Additionally, our current problems break their bottles but this will break new build-from-source builds.

> From that point on that Alias moves to the newest version each update & the "current" version is always the only one that isn't keg_only.

We agree on this: the alias should never be keg_only.

> New versions are introduced without mandating a hoard of revisions or formula depends_on rewrites immediately.

That's the same as my system except in your the `depends_on "boost"` will not be possible again.

> The original name is phased out in favour of an alias over the space of a major release or two.

In some formulae this means the original name won't be phased out for years which means we don't provide version support to these formulae for years.

> brew install <formula> will eventually always install the latest formula, whilst getting there in the least breaking manner.

The "eventually" is a bit concerning here as we're reliant on upstream release processes.

> Once superenv filters HOMEBREW_PREFIX/include and HOMEBREW_PREFIX/lib we stop caring what people choose to link into HOMEBREW_PREFIX in terms of what version is linked publicly and use purely opt_prefixes to link and reference against.

I'm 👍  with this.

> I believe ILZ is loosely on board with this idea going forwards after his work on Boost this past week or so, but he's very welcome to disagree if I misunderstood that.

MikeMcQuaid commented:
@DomT4 So in short the main deviation I see between your plan and mine is that you don't think we should rename the existing `boost` formula. I think we should because the rename support is well tested at this point and provides the necessary fallbacks.

MikeMcQuaid commented:
> This migration requires making everything depends_on "boost@1.6.2" even for stuff that doesn't break on Boost version updates. This feels like a step back rather than forward to me.

Furthermore: this would be a breaking API change. We will break infrequently updated tap formulae and we will likely get the blame for that (justifiably so, I think). We should aim to provide a new API (e.g. `depends_on "boost@1.6.2"`) while allowing the existing API (e.g. `depends_on "boost"`) to work as well/badly as it currently does.

DomT4 commented:
I'm not completely sure you read my comment through in its entirety because some of the stuff you raised conflicts with stuff I'd already outlined above, especially the statement that removing `boost.rb` later down the line will break third-party taps. I explicitly stated in my original comment that removing `boost.rb` would be immediately followed up by adding a `boost` alias pointing to the latest available `boost@x.y` formula at the time, which completely facilitates `depends_on "boost"` being possible.

I'll go ahead and admit my system is full of issues, but I think your's is as well to be honest, such as the built-in system where `boost@1.63.rb` goes in as `keg_only` and then relatively soon after stops being `keg_only`, which mandates a revision bump, so you're likely to give folks a `boost@1.63` installation and then whomp them again fairly quickly with a rebuild. That's not such a huge issue given the `cellar :any`, but increasingly unless we make strides on `cellar :any` that could become punishing to non-`/usr/local` users. It also kind of sucks for users on slower or more limited internet connections than we can afford to enjoy ourselves.

Neither of us have touched on the issue of orphaned formulae, which are going to become quite a bit more common as we go ahead and purge anything outside of +/-2 major versions that is no longer depended upon. `boost` is a good >400MB installation, ending up in a situation where multiple versions are installed and the formulae for some of those are no longer present is less than ideal on disk space & UX. This is already an issue today but it will be a bigger issue as we start mandating adding extra versioned installations of the same formula locally.

But look, life is short and I don't expect a ton of movement here however compelling or awful an argument I piece together, so I'm not going to invest a lot more energy into this. It is what it is.

ilovezfs commented:
>the built-in system where boost@1.63.rb goes in as keg_only and then relatively soon after stops being keg_only, which mandates a revision bump

Is there no way to make that transition without a revision bump?

MikeMcQuaid commented:
> But look, life is short and I don't expect a ton of movement here however compelling or awful an argument I piece together, so I'm not going to invest a lot more energy into this. It is what it is.

I'm happy to wait for and value your feedback on this @DomT4.

MikeMcQuaid commented:
@DomT4 and I talked privately about this and he's 🆗 with this approach being merged as-is.

> Is there no way to make that transition without a revision bump?

If `revision` bumps are required at undocumented points throughout this process I consider that a bug that we should address. `revision`s are a sledgehammer we use a bit too often and should investigate ways to avoid them.

DomT4 commented:
Yeah, I'm going to be around considerably less for at least a while, so whatever Mike and @ilovezfs can hammer out and find agreeable between them is fine with me.

> Is there no way to make that transition without a revision bump?

You'd need a new mechanism for it but I _guess_ you could "teach" `brew` that an alias should always not be `keg_only` and have it `brew unlink` any related `@` formulae and `brew link --force` the alias or something when that alias changes during `brew update`.

The only problem with that is that in the case of `keg_only` formulae like `openssl` you want all of the versions to be `keg_only` regardless of if there's an alias pointing to it or not.
